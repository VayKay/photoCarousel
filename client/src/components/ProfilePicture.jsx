/* eslint-disable class-methods-use-this */
/* eslint-disable no-param-reassign */
/* eslint-disable func-names */
/* eslint-disable jsx-a11y/click-events-have-key-events */
/* eslint-disable jsx-a11y/no-noninteractive-element-interactions */
/* eslint-disable react/prop-types */
/* eslint-disable no-console */
/* eslint-disable react/jsx-first-prop-new-line */
/* eslint-disable react/jsx-max-props-per-line */
import React from 'react';
import styles from './style/ProfilePicture.css';

class ProfilePicture extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      height: Math.ceil((window.innerHeight / 2)),
      width: Math.ceil((window.innerWidth / 2)),
      widthPercent: Math.ceil((this.props.totalWidth * 0.5)),
      hover: false,
    };

    this.progressiveLoading = this.progressiveLoading.bind(this);
    this.updateDimensions = this.updateDimensions.bind(this);
    this.hoverHandler = this.hoverHandler.bind(this);
    this.onClickHandler = this.onClickHandler.bind(this);
  }

  componentDidMount() {
    window.addEventListener('resize', this.updateDimensions);
    this.progressiveLoading();
    this.updateDimensions();
  }

  componentWillUnmount() {
    window.removeEventListener('resize', this.updateDimensions);
  }

  onClickHandler() {
    const { clickHandler } = this.props;
    clickHandler('profile');
  }

  progressiveLoading() {
    const imagesToLoad = document.querySelectorAll('img[data-src]');
    const loadImages = function (image) {
      image.setAttribute('src', image.getAttribute('data-src'));
      image.onload = function () {
        console.log(image.classList);
        image.removeAttribute('data-src');
        image.classList.remove(image.classList[1]);
      };
    };
    imagesToLoad.forEach((img) => {
      loadImages(img);
    });
  }

  updateDimensions() {
    const {
      threshold, lastThreshold, currWidth, totalWidth,
    } = this.props;
    if (currWidth < lastThreshold || totalWidth <= 744) {
      this.setState({
        width: Math.ceil((window.innerWidth / 2)),
        height: Math.ceil((window.innerWidth / 2)),
        widthPercent: '100%',
      });
    } else if ((currWidth < threshold && currWidth > lastThreshold) || totalWidth <= 1128) {
      this.setState({
        width: Math.ceil((window.innerWidth / 2)),
        height: Math.ceil((window.innerWidth / 2)),
        widthPercent: '75%',
      });
    } else {
      this.setState({
        width: Math.ceil((window.innerWidth / 2)),
        height: Math.ceil((window.innerWidth / 2)),
        widthPercent: '50%',
      });
    }
  }

  hoverHandler() {
    const { hover } = this.state;
    const { hoverHandler } = this.props;
    if (!hover) {
      hoverHandler('profile');
    } else {
      hoverHandler('clear');
    }
    this.setState({ hover: !hover });
  }


  render() {
    const { height, width, widthPercent } = this.state;
    const {
      totalHeight, totalWidth, opacity, photo, currWidth, threshold, lastThreshold, clickHandler, tinyPhoto,
    } = this.props;
    return (
      <div className={styles.profileContainer}>
        <div className={styles.imgHoverZoom}>
          <img src={tinyPhoto} data-src={photo} id="profilePicture" className={[styles.picture, styles.blur].join(' ')} alt=""
            style={{
              top: 0, minHeight: Math.ceil((totalHeight * 0.4)), maxHeight: '60%', opacity, background: `url(${tinyPhoto})`,
            }}
            height={Math.ceil((totalHeight * 0.6))} width={widthPercent} onMouseEnter={this.hoverHandler}
            onMouseLeave={this.hoverHandler} onClick={this.onClickHandler} onMouseDown={this.hoverHandler}
          />
        </div>
      </div>
    );
  }
}

export default ProfilePicture;
