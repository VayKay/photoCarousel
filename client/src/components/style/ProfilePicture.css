.imgHoverZoom {
    -webkit-box-sizing: border-box;
    -moz-box-sizing: border-box;
    box-sizing: border-box;
    -ms-flex: 0 1 auto;
}

.imgHoverZoom:hover img {
    transform: scale(1.05) !important;
    transition: -ms-transform 450ms cubic-bezier(0.645, 0.045, 0.355, 1) 0s, 
    -webkit-transform 450ms cubic-bezier(0.645, 0.045, 0.355, 1) 0s, transform 450ms 
    cubic-bezier(0.645, 0.045, 0.355, 1) 0s !important;
}

.profileContainer {
    display: inline-block !important;
    overflow: hidden !important;
    left: 0px !important;
    border: 1px solid;
    position: absolute !important;
    background-color: black !important;
}

.picture {
    transform: scale(1) !important;
    background-size: cover!important;
    background-position: center bottom!important;
    object-fit:cover;
    transition: filter 0.5s, -ms-transform 450ms cubic-bezier(0.645, 0.045, 0.355, 1) 0s, -webkit-transform 450ms cubic-bezier(0.645, 0.045, 0.355, 1) 0s, transform 450ms cubic-bezier(0.645, 0.045, 0.355, 1) 0s, opacity 450ms cubic-bezier(0.645, 0.045, 0.355, 1) 0s !important;
    filter: blur(0em);
}

.blur {
    filter: blur(0.9em);
    opacity: 1 !important;
}