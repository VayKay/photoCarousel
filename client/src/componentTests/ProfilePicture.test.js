/* eslint-disable no-undef */
import React from 'react';
import { shallow } from 'enzyme';
import ProfilePicture from '../components/ProfilePicture';

describe('ProfilePicture', () => {
  it('should render correctly', () => {
    const component = shallow(<ProfilePicture />);

    component.setState({
      currWidth: 1440, lastThreshold: 720, opacity: 1, threshold: 1008, totalHeight: 766, totalWidth: 1440, photo: 'http://lorempixel.com/1440/960/city',
    });

    expect(component).toMatchSnapshot();
  });

  it('should call load profile picture progressively on mount', () => {
    const spy = jest.spyOn(ProfilePicture.prototype, 'progressiveLoading');
    const component = shallow(<ProfilePicture />);
    component.setState({
      currWidth: 1440, lastThreshold: 720, opacity: 1, threshold: 1008, totalHeight: 766, totalWidth: 1440, photo: 'http://lorempixel.com/1440/960/city',
    });

    component.instance().progressiveLoading();
    expect(spy).toHaveBeenCalled();
  });

  it('should handle hover correctly', () => {
    const mockCallBack = jest.fn();
    const component = shallow(<ProfilePicture hoverHandler={mockCallBack} />);

    component.find('img').simulate('mouseenter');
    expect(mockCallBack.mock.calls.length).toEqual(1);
  });

  it('should handle click correctly', () => {
    const mockCallBack = jest.fn();
    const component = shallow(<ProfilePicture clickHandler={mockCallBack} />);

    component.find('img').simulate('click');
    expect(mockCallBack.mock.calls.length).toEqual(1);
  });

  it('should handle window resize', () => {
    const component = shallow(<ProfilePicture />);

    const spy = jest.spyOn(component.instance(), 'updateDimensions');

    global.addEventListener('resize', spy);
    global.dispatchEvent(new Event('resize'));

    expect(spy).toHaveBeenCalled();
  });
});
