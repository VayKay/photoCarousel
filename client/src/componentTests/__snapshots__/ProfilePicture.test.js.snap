// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProfilePicture should render correctly 1`] = `
<div
  className="profileContainer"
>
  <div
    className="imgHoverZoom"
  >
    <img
      alt=""
      className="picture blur"
      height={NaN}
      id="profilePicture"
      onClick={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      style={
        Object {
          "background": "url(undefined)",
          "maxHeight": "60%",
          "minHeight": NaN,
          "opacity": undefined,
          "top": 0,
        }
      }
      width="50%"
    />
  </div>
</div>
`;
